cmd/juju/application/deployer/bundlehandler.go

```diff
@@ -1749,22 +1749,26 @@ func isUserAlreadyHasAccessErr(err error) bool {
 	return err != nil && strings.Contains(err.Error(), "user already has")
 }
 
+func addCharmConstraintsParser(defaultConstraints constraints.Value) func(string) bundlechanges.ArchConstraint {
+	return func(s string) bundlechanges.ArchConstraint {
+		return bundleArchConstraint{
+			constraints:        s,
+			defaultConstraints: defaultConstraints,
+		}
+	}
+}
+
 type bundleArchConstraint struct {
-	constraints string
+	constraints        string
+	defaultConstraints constraints.Value
 }
 
 func (b bundleArchConstraint) Arch() (string, error) {
 	cons, err := constraints.Parse(b.constraints)
 	if err != nil {
 		return "", errors.Trace(err)
 	}
-	return arch.ConstraintArch(cons, nil), nil
-}
-
-func addCharmConstraintsParser(s string) bundlechanges.ArchConstraint {
-	return bundleArchConstraint{
-		constraints: s,
-	}
+	return arch.ConstraintArch(cons, &b.defaultConstraints), nil
```

Comment:
We just need to make sure we pass in the model constraints otherwise it will fall back to the default architecture.